sys/net/network_layer/fib/fib.c

```diff
@@ -609,6 +691,712 @@ int fib_get_num_used_entries(void)
+        LL_FOREACH_SAFE(fib_sr->sr_path, elt, tmp) {
+            universal_address_rem(elt->address);
+            elt->address = NULL;
+            LL_DELETE(fib_sr->sr_path, elt);
+        }
+        fib_sr->sr_path = NULL;
+    }
+
+    mutex_unlock(&mtx_access);
+    return 0;
+}
+
+int fib_sr_next(fib_sr_t *fib_sr, fib_sr_entry_t **sr_path_entry)
+{
+    timex_t now;
+    vtimer_now(&now);
+
+    mutex_lock(&mtx_access);
+    if (fib_sr == NULL || sr_path_entry == NULL) {
+        mutex_unlock(&mtx_access);
+        return -EFAULT;
+    }
+
+    if (fib_sr->sr_path == NULL) {
+        mutex_unlock(&mtx_access);
+        return -EFAULT;
+    }
+
+    if (fib_sr_check_lifetime(fib_sr, &now) == -ENOENT) {
+        mutex_unlock(&mtx_access);
+        return -ENOENT;
+    }
+
+    /* if we reach the destination entry, i.e. the last entry we just return 1 */
+    if (*sr_path_entry == fib_sr->sr_dest) {
+        mutex_unlock(&mtx_access);
+        return 1;
+    }
+
+    /* when we start, we pass the first entry */
+    if (*sr_path_entry == NULL) {
+        *sr_path_entry = fib_sr->sr_path;
+    }
+    else {
+        /* in any other case we just return the next entry */
+        *sr_path_entry = (*sr_path_entry)->next;
+    }
+
+    mutex_unlock(&mtx_access);
+    return 0;
+}
+
+int fib_sr_search(fib_sr_t *fib_sr, uint8_t *addr, size_t addr_size,
+                  fib_sr_entry_t **sr_path_entry)
+{
+    timex_t now;
+    vtimer_now(&now);
+
+    mutex_lock(&mtx_access);
+    if (fib_sr == NULL || addr == NULL || sr_path_entry == NULL) {
+        mutex_unlock(&mtx_access);
+        return -EFAULT;
+    }
+
+    if (fib_sr_check_lifetime(fib_sr, &now) == -ENOENT) {
+        mutex_unlock(&mtx_access);
+        return -ENOENT;
+    }
+
+    fib_sr_entry_t *elt;
+    LL_FOREACH(fib_sr->sr_path, elt) {
+        size_t addr_size_match = addr_size << 3;
+        if (universal_address_compare(elt->address, addr, &addr_size_match) == 0) {
+
+            /* temporary workaround to calm compiler */
+            (void)sr_path_entry;
+
+            *sr_path_entry = elt;
+            mutex_unlock(&mtx_access);
+            return 0;
+        }
+    }
+
+    mutex_unlock(&mtx_access);
+    return -EHOSTUNREACH;
+}
+
+int fib_sr_entry_append(fib_sr_t *fib_sr, uint8_t *addr, size_t addr_size)
+{
+    timex_t now;
+    vtimer_now(&now);
+
+    mutex_lock(&mtx_access);
+    if (fib_sr == NULL || addr == NULL) {
+        mutex_unlock(&mtx_access);
+        return -EFAULT;
+    }
+
+    if (fib_sr_check_lifetime(fib_sr, &now) == -ENOENT) {
+        mutex_unlock(&mtx_access);
+        return -ENOENT;
+    }
+
+    fib_sr_entry_t *elt;
+    LL_FOREACH(fib_sr->sr_path, elt) {
+        size_t addr_size_match = addr_size << 3;
+        if (universal_address_compare(elt->address, addr, &addr_size_match) == 0) {
+
+            mutex_unlock(&mtx_access);
+            return -EINVAL;
+        }
+    }
+
+    fib_sr_entry_t *new_entry[1];
```

Comment:
I need a "pointer pointer" here to manipulate the inner pointer within the following line +ff:
``` C
fib_sr_new_entry(addr, addr_size, &new_entry[0]);
```